ddl/ddl_api.go

```diff
@@ -1104,6 +1104,47 @@ func checkConstraintNames(constraints []*ast.Constraint) error {
 	return nil
 }
 
+// checkInvisibleIndexOnPK check if primary key is invisible index.
+func checkInvisibleIndexOnPK(tblInfo *model.TableInfo) error {
+	pk := getPrimaryKey(tblInfo)
+	if pk != nil && pk.Invisible {
+		return ErrPKIndexCantBeInvisible
+	}
+	return nil
+}
+
+// getPrimaryKey extract the primary key in a table and return `IndexInfo`
+// The returned primary key could be explicit or implicit.
+// If there is no explicit primary key in table,
+// the first UNIQUE INDEX on NOT NULL columns will be the implicit primary key.
+// For more information about implicit primary key, see
+// https://dev.mysql.com/doc/refman/8.0/en/invisible-indexes.html
+func getPrimaryKey(tblInfo *model.TableInfo) *model.IndexInfo {
+	var implicitPK *model.IndexInfo
+
+	for _, key := range tblInfo.Indices {
+		if key.Primary {
+			// table has explicit primary key
+			return key
+		}
+		// find the first unique key with NOT NULL columns
+		if implicitPK == nil && key.Unique {
+			// ensure all columns in unique key have NOT NULL flag
+			allColNotNull := true
+			for _, idxCol := range key.Columns {
+				col := model.FindColumnInfo(tblInfo.Cols(), idxCol.Name.L)
+				if !mysql.HasNotNullFlag(col.Flag) {
+					allColNotNull = false
+				}
+			}
+			if allColNotNull {
+				implicitPK = key
```

Comment:
@bb7133 No we need return the `PK` in this function, because L1110-L1112 need this `indexInfo` to check whether this `PK` is invisible.

Reply:
What I mean is `return key` here. I think it should be fine.

Reply:
But in this case,
```sql
create table t(a int NOT NULL, b int primary key, unique(a));
```
The index `a` become implicit PK in 1st iterator, but we cannot return `a` directly here, because PK on `b` becomes explicit PK in the 2nd iterator, which make index `a` no longer be implicit PK (If a table have an explicit PK, no one will become implicit PK).
So we must iterator all indexes on this table, and then return the implicit PK.

Reply:
I see, thanks!